### PR TITLE
Limit  `page_size` to 20 for unauthenticated requests

### LINKED ISSUE #715: Ingestion of images stops unexpectedly after cleanup
## Description

All CI flakiness has boiled down to the ingestion of images stopping unexpectedly after the cleanup step. Rerunning ingestion almost surely fixes the issue.

We should investigate the issue and find out the cause. If it is incurable, we should add a step to retry the ingestion which is almost guaranteed to succeed the second time.

This also causes #562.

## Reproduction
See failed runs:
- https://github.com/WordPress/openverse-api/runs/6527207725?check_suite_focus=true#step:8:702
- https://github.com/WordPress/openverse-api/runs/6527646327?check_suite_focus=true#step:11:480

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

### LINKED ISSUE #751: Limit unauthenticated API requests to 20 page size
## Problem

Currently it's possible to send page_size up to 500. This makes it trivial for someone to DDoS the API, even if they're getting throttled.

Note: I'm marking this high instead of critical because we can't deploy the API right now anyway to fix it (although maybe this would help prevent the API issues we were seeing, something to investigate).

## Description

Limit the page_size to 20 for requests without authorization to do otherwise (which currently is none!)

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by krysal**
## Fixes

Fixes #751 by @sarayourfriend 

## Description

Adds the restriction for list views.

## Testing Instructions

Make authenticated and unauthenticated requests to try the new limit.

```sh
curl -s "http://localhost:8000/v1/images/?page_size=22"
```

You should get: `{
	"detail": "Page size must be between 0 & 20 for unauthenticated requests."
}` and for

```sh
curl -H "Authorization: Bearer <access_token>" \
  "http://localhost:8000/v1/images/?page_size=22"
```
you should get results normally.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/catalog/api/views/media_views.py [hunk 1]
```diff
@@ -52,7 +52,17 @@ def get_queryset(self):
...
-        self.paginator.page_size = request.query_params.get("page_size")
+        raw_page_size = request.query_params.get("page_size")
+        if (
+            request.user
+            and request.user.is_anonymous
+            and raw_page_size is not None
+            and int(raw_page_size) > 20
+        ):
+            raise get_api_exception(
+                "Page size must be between 0 & 20 for unauthenticated requests.", 400
```
Comment (sarayourfriend) [inline api/catalog/api/views/media_views.py:63]:
```suggestion
                "Page size must be between 0 & 20 for unauthenticated requests.", 401
```

401 is Unauthorized, which I think is more accurate here than "bad request" :slightly_smiling_face:

**Review by sarayourfriend — COMMENTED** [review]
This approach looks good to me :slightly_smiling_face:

**Review by AetherUnbound — COMMENTED** [review]
I love this approach! Initially I was thinking about just reducing any larger page size to 20, but sending a 401 back is an excellent idea 💯

File: api/test/auth_test.py [hunk 2]
```diff
@@ -80,3 +78,25 @@ def test_auth_rate_limit_reporting(client, test_auth_token_exchange, verified=Fa
...
+@pytest.mark.django_db
+def test_page_size_limit_authed(client, test_auth_token_exchange):
+    time.sleep(1)
+    token = test_auth_token_exchange["access_token"]
+    query_params = {"filter_dead": False, "page_size": 21}
+    res = client.get("/v1/images/", query_params, HTTP_AUTHORIZATION=f"Bearer {token}")
+    assert res.status_code == 200
+
+    query_params = {"filter_dead": False, "page_size": 500}
+    res = client.get("/v1/images/", query_params, HTTP_AUTHORIZATION=f"Bearer {token}")
```
[the remark this answers is not in the document]
Comment (krysal) [reply api/test/auth_test.py:102]: Oops! That was the intention! Thank you for pointing it out. Adding it.

**Review by dhruvkb — COMMENTED** [review]
Can we add this validation inside the serializer using `validate_page_size`? That would be better imo as it keeps all our validation localised in one place.

**Comment by krysal** [comment]
Thank you, @dhruvkb! I was looking for something like that, it is much cleaner this way ✨

File: api/catalog/api/serializers/media_serializers.py [hunk 3]
```diff
@@ -149,6 +156,14 @@ def validate_tags(self, value):
     def validate_title(self, value):
         return self._truncate(value)
 
+    def validate_page_size(self, value):
+        request = self.context.get("request", None)
+        if request.user and request.user.is_anonymous and value > 20:
+            raise get_api_exception(
+                "Page size must be between 1 & 20 for unauthenticated requests.", 401
+            )
+        return value
```
Comment (sarayourfriend) [inline api/catalog/api/serializers/media_serializers.py:165]: The `None` default is redundant on dictionary's `get` method, but more importantly we need to actually handle the None case here or we'll end up with 500s:

```suggestion
        request = self.context.get("request", None)
        authenticated = request and request.user and not request.user.is_anonymous
        if not authenticated and value > 20:
            raise get_api_exception(
                "Page size must be between 1 & 20 for unauthenticated requests.", 401
            )
        return value
```

That seems like a reasonable and safe way to handle None. What you do think? Just treat it the same as an anonymous request?

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
LGTM! Dhruv's suggestion makes it way cleaner, kudos @dhruvkb! Just one edge case we should handle but otherwise this is g2g for me :slightly_smiling_face:

File: api/catalog/api/serializers/media_serializers.py (hunk 3 above)
Comment (krysal) [reply api/catalog/api/serializers/media_serializers.py]: I don't think that case will happen but I like this! Though it's a bit tricky to check using the condition for the authenticated user (it becomes `None` for some reason 🤷🏻‍♀️). I modified it a bit, LMKWYT.

**Review by AetherUnbound — APPROVED** [review]
Looks great! Great suggestion to add this to page size validation @dhruvkb 💯

File: api/catalog/api/serializers/media_serializers.py (hunk 3 above)
Comment (sarayourfriend) [reply api/catalog/api/serializers/media_serializers.py]: Looks good to me! I also don't foresee it happening but "something something [Murphy's Law](https://en.wikipedia.org/wiki/Murphy%27s_law)" :sweat_smile:

**Review by sarayourfriend — APPROVED** [review]

**Pull request merged**